### PR TITLE
Support login to the integrated registry via CLI

### PR DESCRIPTION
**Pull request by smarterclayton**
Adds a new `oc image login` command that allows the user to easily log
in to the integrated registry. By default it tries to use the current
user token (but supports `-z` for service accounts) and looks up the
public or internal hostname by finding image streams in the current
namespace or the 'openshift' namespace (which is usually populated).

This did highlight that openshift ansible is not setting externalRegistryHostname by default in the install which it should be doing.

### CODE REVIEW

**Comment by smarterclayton** [comment]
@bparees for 90% of clusters this will do the right thing once we start forcing registryExternalHostname, since everyone is going to have stuff in openshift.  Eventually when david's changes for logging /metrics route publishing to config map land we can also push the registry URL.

**Comment by bparees** [comment]
i don't suppose i could convince you that it should be "oc registry" and not "oc image"?

**Comment by smarterclayton** [comment]
if we have commands that work on both images and registries would we want
two top level commands?

oc image mirror
oc image rebase
oc image login
oc image import
oc image add-layer

oc registry mirror
oc registry rebase
oc registry login
oc registry import
oc registry add-layer

?  Agree registry sounds better for some, but I'd also prefer for a top
level command to have more related subcommands (i.e. two almost empty top
level commands makes me sad).

Agree registry will be more discoverable for some.

**Comment by smarterclayton** [comment]
this is updated with feedback and adds a `registry info` command that functions a bit like `whoami` for scripting interactions with the server.

File: pkg/oc/cli/cmd/registry/login/login.go [hunk 1]
```diff
@@ -0,0 +1,309 @@
...
+		ns, _, err := f.DefaultNamespace()
+		if err != nil {
+			return err
+		}
+
+		if registry, internal := findPublicHostname(client, ns, "openshift"); len(registry) > 0 {
+			if ref, err := imageapi.ParseDockerImageReference(registry); err == nil {
+				o.HostPort = ref.Registry
+				if internal {
+					fmt.Fprintf(o.ErrOut, "info: Using internal registry hostname %s\n", o.HostPort)
```
Comment (soltysh) [inline pkg/oc/cli/cmd/registry/login/login.go]: If this is info, use `o.Out` not `o.ErrOut`, the latter is only for warning and errors.

Comment (smarterclayton) [reply pkg/oc/cli/cmd/registry/login/login.go]: We should be using stderr for anything that is not directly tied to the output of the command.

Comment (smarterclayton) [reply pkg/oc/cli/cmd/registry/login/login.go]: Info, warning, and error are all stderr.

Comment (soltysh) [reply pkg/oc/cli/cmd/registry/login/login.go]: Warning and error -> stderr, if info should be possible to hide use glog with different levels. Any other info is info that should put to stdout just like the rest of the output b/c it's relevant to the user.

Comment (smarterclayton) [reply pkg/oc/cli/cmd/registry/login/login.go]: No, it shouldn’t.  Info is not glog.  Info is information relevant to the user that is not a warning.  It is not normal output.  Stdout is data for processing.  Glog is not human facing the way info is.

Comment (smarterclayton) [reply pkg/oc/cli/cmd/registry/login/login.go]:
```
grep -R --include=*.go -E "\"info:" pkg/
pkg//oc/admin/migrate/etcd/ttl.go:	fmt.Fprintf(o.Out, "info: Lease #%d with TTL %d created\n", lease.ID, lease.TTL)
pkg//oc/admin/migrate/etcd/ttl.go:	fmt.Fprintf(o.Out, "info: Attaching lease to %d entries\n", len(objectsResp.Kvs))
pkg//oc/admin/migrate/etcd/ttl.go:		fmt.Fprintf(o.Out, "info: Lease already attached to %d entries, no change made\n", alreadyAttached)
pkg//oc/admin/migrate/migrator.go:		fmt.Fprintf(out, "info: to rerun only failing resources, add --include=%s\n", strings.Join(t.resourcesWithErrors.List(), ","))
pkg//oc/admin/router/router.go:			fmt.Fprintf(errout, "info: password for stats user %s has been set to %s\n", cfg.StatsUsername, cfg.StatsPassword)
pkg//oc/cli/cmd/image/mirror/mirror.go:				fmt.Fprintf(o.ErrOut, "info: Filtered all images from %s, skipping\n", src.ref)
pkg//oc/cli/cmd/logs.go:		o.KubeLogOptions.Out.Write([]byte(fmt.Sprintf("info: Logs available at %s\n", urlString)))
pkg//oc/cli/cmd/set/buildhook.go:			fmt.Fprintf(o.Err, "info: %s %q was not changed\n", info.Mapping.Resource, info.Name)
pkg//oc/cli/cmd/set/buildsecret.go:			fmt.Fprintf(o.Err, "info: %s %q was not changed\n", info.Mapping.Resource, info.Name)
pkg//oc/cli/cmd/set/deploymenthook.go:			fmt.Fprintf(o.Err, "info: %s %q was not changed\n", info.Mapping.Resource, info.Name)
pkg//oc/cli/cmd/set/imagelookup.go:			fmt.Fprintf(o.Err, "info: %s %q was not changed\n", info.Mapping.Resource, info.Name)
pkg//oc/cli/cmd/set/probe.go:			fmt.Fprintf(o.Err, "info: %s %q was not changed\n", info.Mapping.Resource, info.Name)
pkg//oc/cli/cmd/set/routebackends.go:			fmt.Fprintf(o.Err, "info: %s %q was not changed\n", info.Mapping.Resource, info.Name)
pkg//oc/cli/cmd/set/triggers.go:			fmt.Fprintf(o.Err, "info: %s %q was not changed\n", info.Mapping.Resource, info.Name)
pkg//oc/cli/cmd/set/volume.go:			fmt.Fprintf(v.Err, "info: %s %q was not changed\n", info.Mapping.Resource, info.Name)
pkg//oc/cli/cmd/set/volume.go:		fmt.Fprintf(v.Err, "info: Generated volume name: %s\n", name)
pkg//oc/cli/util/clientcmd/factory.go:		glog.V(1).Infof("info: the output version specified is invalid. %s\n", defaultVersionInfo)
```

Comment (smarterclayton) [reply pkg/oc/cli/cmd/registry/login/login.go]: All glog output goes to stderr, but `info:` is effectively v(0) with the prefix (which is not useful for a human interacting with a cli).  The user *should* have the information.  The user *should not* have to filter that out, i.e.:

```
$ curl https://$( oc registry info )
```

shouldn't send `info: The registry is internal` to curl - it should print that on stderr so the shell captures only the url, not the output

Comment (smarterclayton) [reply pkg/oc/cli/cmd/registry/login/login.go]: The TTL command is outputting to stdout because it prints nothing else.

**Comment by smarterclayton** [comment]
If no other comments marking.

**Comment by bparees** [comment]
nothing from me

File: pkg/oc/cli/cmd/registry/info/info.go [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+	kcmdutil "k8s.io/kubernetes/pkg/kubectl/cmd/util"
+
+	imageclient "github.com/openshift/client-go/image/clientset/versioned"
+	imageapi "github.com/openshift/origin/pkg/image/apis/image"
+	"github.com/openshift/origin/pkg/image/registryclient"
+	"github.com/openshift/origin/pkg/oc/cli/util/clientcmd"
+)
+
+var (
+	loginDesc = templates.LongDesc(`
```
Comment (bparees) [inline pkg/oc/cli/cmd/registry/info/info.go]: infoDesc

File: pkg/oc/cli/cmd/registry/info/info.go [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+		Display information about the integrated registry
+
+		This command exposes information about the integrated registry, if configured.
+		Use --check to verify your local client can access the registry. If the adminstrator
+		has not configured a public hostname for the registry then this command may fail when
+		run outside of the server.
+
+		Experimental: This command is under active development and may change without notice.`)
+
+	loginExample = templates.Examples(`
```
Comment (bparees) [inline pkg/oc/cli/cmd/registry/info/info.go]: infoExample

File: pkg/oc/cli/cmd/registry/info/info.go [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+	o := &Options{
+		Out:    out,
+		ErrOut: errOut,
+	}
+
+	cmd := &cobra.Command{
+		Use:     "info ",
+		Short:   "Print info about the integrated registry",
+		Long:    loginDesc,
+		Example: fmt.Sprintf(loginExample, name+" login"),
```
Comment (bparees) [inline pkg/oc/cli/cmd/registry/info/info.go]: infoExample

File: pkg/oc/cli/cmd/registry/info/info.go [hunk 5]
```diff
@@ -0,0 +1,196 @@
...
+func New(name string, f *clientcmd.Factory, out, errOut io.Writer) *cobra.Command {
+	o := &Options{
+		Out:    out,
+		ErrOut: errOut,
+	}
+
+	cmd := &cobra.Command{
+		Use:     "info ",
+		Short:   "Print info about the integrated registry",
+		Long:    loginDesc,
```
Comment (bparees) [inline pkg/oc/cli/cmd/registry/info/info.go]: infoDesc

**Pull request merged**